### PR TITLE
Jennyf/user credential tests

### PR DESCRIPTION
**Pull request by jennyf19**
#852 
Adding further test coverage for federated user

### CODE REVIEW

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 1]
```diff
@@ -1,11 +1,15 @@
 ﻿using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal;
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: add MIT header

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 2]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
             // All mocks are consumed
             Assert.AreEqual(0, HttpMessageHandlerFactory.MockHandlersCount());
         }
+
+        [TestMethod]
+        [Description("Provide only username of Federated user. Result in integrated auth call with no password xml tag.")]
+        public async Task ProvideUsernameOfFederatedUser_NoPasswordXmlTagTestAsync()
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: rename the test to what it is and what it does. say, something like IntegratedAuthUsingUpn_GetsTokenFromService()

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 3]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+                }
+            });
+
+            // Call acquire token
+            AuthenticationResult result = await context.AcquireTokenAsync(TestConstants.DefaultResource, TestConstants.DefaultClientId,
+                new UserCredential(TestConstants.DefaultDisplayableId));
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual("some-access-token", result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultDisplayableId, result.UserInfo.DisplayableId);
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: check that cache actually has the entry

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 4]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+                    {"api-version", "1.0"}
+                }
+            });
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler("https://msft.sts.microsoft.com/adfs/services/trust/mex")
+            {
+                Method = HttpMethod.Get,
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.NotFound)
+                {
+                    Content = new StringContent(File.ReadAllText("TestMex.xml"))
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: 404 should not be returning XML content. It should be "Not found"

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 5]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+            Assert.AreEqual(TestConstants.DefaultDisplayableId, result.UserInfo.DisplayableId);
+
+            // All mocks are consumed
+            Assert.AreEqual(0, HttpMessageHandlerFactory.MockHandlersCount());
+        }
+
+
+        [TestMethod]
+        [Description("No password provided and Mex does not return integrated auth endpoint.")]
+        public async Task NoPasswordAndMexEndpointFailsToResolveTestAsync()
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: rename to IntegratedAuthMissingMex_FailsMexParsing

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 6]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+            {
+                Method = HttpMethod.Get,
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.NotFound)
+                {
+                    Content = new StringContent(File.ReadAllText("TestMex.xml"))
+                }
+            });
+
+            string emptyPassword = null;
+            UserPasswordCredential userPasswordCredential = new UserPasswordCredential(TestConstants.DefaultDisplayableId, emptyPassword);
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: no password means, you should be using UserCredential class.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 7]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+            // Check inner exception
+            Assert.AreEqual(" Response status code does not indicate success: 404 (NotFound).", result.InnerException.Message);
+
+            // All mocks are consumed
+            Assert.AreEqual(0, HttpMessageHandlerFactory.MockHandlersCount());
+        }
+
+        [TestMethod]
+        [Description("Test for AcquireToken with valid token in cache")]
+        public async Task AcquireTokenWithValidTokenInCache_ReturnsCachedTokenTestAsync()
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: this is a managed user test case form the code. You should be using UserCredential class and mock call should be failing with 400 saying "no network call is expected". this is to ensure that no network call actually gets made and if it does then we do know how to fix it

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs [hunk 8]
```diff
@@ -89,5 +94,208 @@ public void Initialize()
...
+            Assert.AreEqual("existing-access-token", result.AccessToken);
+
+            // There should be one cached entry.
+            Assert.AreEqual(1, context.TokenCache.Count);
+            Assert.IsNotNull(result.UserInfo);
+        }
+
+        [TestMethod]
+        [Description("Mex endpoint fails to resolve and results in a 404")]
+        public async Task MexEndpointFailsToResolveTestAsync()
```
Comment (kpanwar) [inline tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: how is this test different than the one you wrote above for mex failure? you should repurpose this test to return a malformed XML and make sure XML parser fails.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 1 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: done.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 2 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: done.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 5 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: done.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 4 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: done.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 6 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: done.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 7 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: Thanks for the explanation on this.

File: tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs (hunk 8 above)
Comment (jennyf19) [reply tests/Test.ADAL.NET.Integration/FederatedUserCredentialTests.cs]: This makes more sense...thanks

**Pull request merged**